src/core/vector/qgsvectorlayerprofilegenerator.cpp

```diff
@@ -802,423 +814,451 @@ bool QgsVectorLayerProfileGenerator::generateProfileForPoints()
-          std::unique_ptr< QgsLineString > extrudedDistanceVHeightRing = std::make_unique< QgsLineString >( newDistance, extrudedZ );
-          std::unique_ptr< QgsLineString > reversedDistanceVHeightExtrusion( extrudedDistanceVHeightRing->reversed() );
-          distanceVHeightRing->append( reversedDistanceVHeightExtrusion.get() );
-          distanceVHeightRing->close();
-          resultFeature.crossSectionGeometry = QgsGeometry( new QgsPolygon( distanceVHeightRing.release() ) );
-        }
-        else
-        {
-          resultFeature.geometry = QgsGeometry( new QgsLineString( newX, newY, newZ ) );
-          resultFeature.crossSectionGeometry = QgsGeometry( new QgsLineString( newDistance, newZ ) );
-        }
-        mResults->features[resultFeature.featureId].append( resultFeature );
+    QgsPoint interpolatedPoint( x, y, z ); // general case (not a triangle)
+
+    *outX++ = x;
+    *outY++ = y;
+    if ( triangle->exteriorRing()->numPoints() <= 4 ) // triangle case
+    {
+      QgsPoint tmpPt = interpolatePointOnTriangle( triangle, x, y );
+      if ( ! tmpPt.isEmpty() ) // point x,y inside the triangle
+      {
+        interpolatedPoint = tmpPt;
       }
     }
-  };
+    *outZ++ = std::isnan( interpolatedPoint.z() ) ? 0.0 : interpolatedPoint.z();
 
-  QgsFeature feature;
-  QgsFeatureIterator it = mSource->getFeatures( request );
-  while ( it.nextFeature( feature ) )
-  {
-    if ( mFeedback->isCanceled() )
-      return false;
+    if ( mExtrusionEnabled )
+      *extZOut++ = ( std::isnan( interpolatedPoint.z() ) ? 0.0 : interpolatedPoint.z() ) + extrusion;
 
-    if ( !mProfileCurveEngine->intersects( feature.geometry().constGet() ) )
-      continue;
+    mResults->mRawPoints.append( interpolatedPoint );
+    mResults->minZ = std::min( mResults->minZ, interpolatedPoint.z() );
+    mResults->maxZ = std::max( mResults->maxZ, interpolatedPoint.z() );
+    if ( mExtrusionEnabled )
+    {
+      mResults->minZ = std::min( mResults->minZ, interpolatedPoint.z() + extrusion );
+      mResults->maxZ = std::max( mResults->maxZ, interpolatedPoint.z() + extrusion );
+    }
 
-    mExpressionContext.setFeature( feature );
+    const double distance = mProfileCurveEngine->lineLocatePoint( interpolatedPoint, &lastError );
+    *outDistance++ = distance;
 
-    const QgsGeometry g = feature.geometry();
-    if ( g.isMultipart() )
-    {
-      for ( auto it = g.const_parts_begin(); it != g.const_parts_end(); ++it )
-      {
-        if ( !mProfileCurveEngine->intersects( *it ) )
-          continue;
+    mResults->mDistanceToHeightMap.insert( distance, interpolatedPoint.z() );
+    lastDistanceAlongProfileCurve = distance;
+  }
 
-        processCurve( feature, qgsgeometry_cast< const QgsCurve * >( *it ) );
-      }
+  // insert nan point to end the line
+  mResults->mDistanceToHeightMap.insert( lastDistanceAlongProfileCurve + 0.000001, qQNaN() );
```

Comment:
```suggestion
  mResults->mDistanceToHeightMap.insert( lastDistanceAlongProfileCurve + 0.000001, qQNaN() );
```

Use std nan methods, not qt's additions (which are highly likely to disappear in a future Qt release! :laughing: )

Reply:
fixed!